storage/src/adb/keyless.rs

```diff
@@ -0,0 +1,863 @@
+            VConfig {
+                partition: cfg.values_journal_partition,
+                compression: cfg.values_compression,
+                codec_config: cfg.values_codec_config,
+                buffer_pool: cfg.buffer_pool,
+                write_buffer: cfg.values_write_buffer,
+            },
+        )
+        .await?;
+
+        // Rewind to the last commit point if necessary.
+        let mut op_index = locations_size;
+        // The most recent commit point (if any).
+        let mut last_commit_loc = None;
+        // The location of the first append operation to follow the last known commit, and the
+        // offset it wraps.
+        let mut rewind_point = None;
+        while op_index > 0 {
+            op_index -= 1;
+            match locations.read(op_index).await? {
+                Operation::Commit => {
+                    last_commit_loc = Some(op_index);
+                    break;
+                }
+                Operation::Append(offset) => {
+                    rewind_point = Some((op_index, offset));
+                }
+            }
+        }
+        if let Some(last_commit_loc) = last_commit_loc {
+            if last_commit_loc != locations_size - 1 {
+                // There's at least one append operation to rewind.
+                warn!(
+                    old_size = locations_size,
+                    new_size = last_commit_loc + 1,
+                    "rewinding to last commit point"
+                );
+                locations.rewind(last_commit_loc + 1).await?;
+                mmr.pop((locations_size - last_commit_loc - 1) as usize)
+                    .await?;
+                locations_size = last_commit_loc + 1;
+                // Rewind the values journal last to ensure the locations journal always references
+                // valid data in the event of failures.
+                let rewind_point = rewind_point.expect("no rewind point found");
+                let section = rewind_point.0 / cfg.values_items_per_section.get();
+                values.rewind_to_offset(section, rewind_point.1).await?;
+            }
+        } else if locations_size > 0 {
+            warn!(
+                old_size = locations_size,
+                "no commit point found, rewinding to start"
+            );
+            locations.rewind(0).await?;
+            mmr.pop(locations_size as usize).await?;
+            locations_size = 0;
+            values.rewind_section(0, 0).await?;
+        }
+
+        Ok(Self {
+            mmr,
+            values,
+            size: locations_size,
+            locations,
+            values_per_section: cfg.values_items_per_section.get(),
+            hasher,
+        })
+    }
+
+    /// Get the value at location `loc` in the database. Returns None if the location is valid but
+    /// does not correspond to an append.
+    pub async fn get(&self, loc: u64) -> Result<Option<V>, Error> {
+        match self.locations.read(loc).await {
```

Comment:
I think you could do `.await?` to avoid the error case (and needing to match here)?